Add capacity to skip H1 Headers

skyetim opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/master/CONTRIBUTING.md
-->

## References

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->
This PR addresses jupyterlab/jupyterlab-toc#66
<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

<!-- Describe the code changes and how they address the issue. -->
There are three main code changes:
1. TOC setting in Advanced Setting Editor is brought back to accommodate the option of skipping h1 headers
2. A new parameter `setting` is added to `createMarkdownGenerator`, `createRenderedMarkdownGenerator` and `createNotebookGenerator` such that the user setting can be passed
3. A new parameter `numberingH1` is added to `generateNumbering` such that it will render correctly if the user decides to skip h1 header

## User-facing changes

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots here. -->

### Notebook
Before
<img width="1680" alt="nb-before" src="https://user-images.githubusercontent.com/16680640/112239374-be067800-8c1c-11eb-9540-ca5922dbd610.png">
After
<img width="1680" alt="nb-after" src="https://user-images.githubusercontent.com/16680640/112239392-c363c280-8c1c-11eb-85af-4984f37a590f.png">

### Markdown
Before
<img width="1680" alt="md-before" src="https://user-images.githubusercontent.com/16680640/112239415-cced2a80-8c1c-11eb-9693-eb94a32e14c2.png">
After
<img width="1680" alt="md-after" src="https://user-images.githubusercontent.com/16680640/112239428-d080b180-8c1c-11eb-84d1-5f9f21087b8a.png">

### Settings
<img width="1680" alt="setting" src="https://user-images.githubusercontent.com/16680640/112239438-d5456580-8c1c-11eb-9e5c-21824db23b3b.png">


## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->
None

skyetim commented:
I don't think the error is on my side. Please advise on next steps.

fcollonval reviewed (changes requested):
Thanks @skyetim 

Overall this looks good. I added some comments and suggestion. Could you also:
- change the code to react to the settings modification so that changing the option does not require refreshing the webpage? - you can use `settings.changed` signal for that (see [example](https://github.com/jupyterlab/extension-examples/blob/8bc09d2839efed8078c9fbc43da0fd39a0b7d0a3/settings/src/index.ts#L55))
- update the code documentation (there are a couple of function for which the input arguments have changed)?

skyetim commented:
> Thanks @skyetim
> 
> Overall this looks good. I added some comments and suggestion. Could you also:
> 
> * change the code to react to the settings modification so that changing the option does not require refreshing the webpage? - you can use `settings.changed` signal for that (see [example](https://github.com/jupyterlab/extension-examples/blob/8bc09d2839efed8078c9fbc43da0fd39a0b7d0a3/settings/src/index.ts#L55))
> * update the code documentation (there are a couple of function for which the input arguments have changed)?

Documentation is updated. 

I am struggling to write the right callback function, as changing the setting requires resetting all generators. What I am thinking is reloading the app, but I am not sure what's the right method to call. Could you give me some pointers on this?

krassowski commented on packages/toc-extension/schema/plugin.json:
```diff
@@ -0,0 +1,14 @@
+  "properties": {
+    "numberingH1": {
+      "title": "Enable h1 numbering",
+      "description": "Specifies whether to number first level headers",
```
```suggestion
      "description": "Whether to number first level headings",
```

These are called headings, headers are a different thing (https://html.spec.whatwg.org/multipage/sections.html)

krassowski commented:
An integrity update is needed for the tests to work (no worries it's not you - just how the things are when adding a dependency):

> Please run `jlpm run integrity` locally and commit the changes

As for the settings question:

> I am struggling to write the right callback function, as changing the setting requires resetting all generators.

A pattern that I use when the changes would normally need to be propagated deep down is to instead of passing the value pass the `settings` object and create a getter like

```ts
get someSetting() {
   return this.setting.composite.someSetting
}
```

and then somewhere else I connect to the `this.settings.changed` signal, roughly:

```
this.settings.changed.connect(() => { this.widget.update()  })
```

but this might be an anti-pattern (I honestly don't know). I believe @fcollonval may have better suggestions.

fcollonval commented on package.json:
```diff
@@ -93,7 +93,11 @@
     }
   },
-  "dependencies": {},
+  "dependencies": {
+    "dependency-graph": "^0.11.0",
+    "sort-package-json": "^1.50.0",
+    "y-codemirror": "^2.1.1"
+  },
```
This is weird. No dependencies should be added here.

fcollonval commented on packages/toc/src/generators/notebook/options_manager.ts:
```diff
@@ -103,6 +109,18 @@ class OptionsManager extends Registry.IOptionsManager {
+   */
+  set numberingH1(value: boolean) {
+    this._numberingH1 = value;
+    this._widget.update();
```
Please condition the update on a change.

fcollonval commented on packages/toc/src/generators/markdown/options_manager.ts:
```diff
@@ -66,6 +72,18 @@ class OptionsManager extends Registry.IOptionsManager {
+   * Gets/sets ToC generator numbering h1 headers.
+   */
+  set numberingH1(value: boolean) {
+    this._numberingH1 = value;
+    this._widget.update();
```
Please condition the update on a change

```suggestion
    if (this._numberingH1 !== value) {
      this._numberingH1 = value;
      this._widget.update();
    }
```

fcollonval reviewed (changes requested):
Hey @skyetim  thanks for pushing this PR.

I like the way you did it with the setter. So to react to settings changes, I would connect to the settings.changed signal in `create...Generator` functions; like it is done for the notebook:

https://github.com/jupyterlab/jupyterlab/blob/64e600e7a53466f9692afe9c3ce0bafbad4ad394/packages/toc/src/generators/notebook/index.ts#L58

But  the callback function would look like:

```ts
settings.changed.connect(() => {options.numberingH1 = settings.composite.numberingH1 as boolean;});
```
Then the update will be triggered from the update call in the setter.

> We are almost there :wink:

fcollonval commented on packages/toc/src/generators/markdown/index.ts:
```diff
@@ -66,19 +80,31 @@ function generate(editor: IDocumentWidget<FileEditor>): INumberedHeading[] {
+  if (settings) {
+    numberingH1 = settings.composite.numberingH1 as boolean;
+  }
+  let options = new OptionsManager(widget, {
```
```suggestion
  const options = new OptionsManager(widget, {
```
`options` is still constant as you modify an attribute of the object not the object itself.

fcollonval commented on packages/toc/src/generators/notebook/index.ts:
```diff
@@ -34,19 +35,31 @@ import { toolbar } from './toolbar_generator';
+  if (settings) {
+    numberingH1 = settings.composite.numberingH1 as boolean;
+  }
+  let options = new OptionsManager(widget, tracker, {
```
```suggestion
  const options = new OptionsManager(widget, tracker, {
```

fcollonval reviewed (commented):
Latest small corrections. Then it is good for me.

krassowski commented on packages/toc/src/generators/notebook/toolbar_generator.tsx:
```diff
@@ -95,6 +95,7 @@ function toolbar(options: OptionsManager, tracker: INotebookTracker) {
             ) as boolean;
             options.initializeOptions(
               numbering || options.numbering,
+              options.numberingH1,
```
Should maybe it possibly be preserved in the notebook metadata as well, so that the setting is remembered when user refreshes the page/re-opens the notebook later? All other toggles work this way. It would require adding:

```
const numberingH1 = tracker.currentWidget.model!.metadata.get(
   'toc-numberingH1'
) as boolean;
```

changing this line to:

```
numberingH1 || options.numberingH1,
```

And updating the setter to also have `this.notebookMetadata = ['toc-numberingH1', this._numberingH1];` as do other setters:

https://github.com/jupyterlab/jupyterlab/blob/dbdefed9db9332381fe4104bdf53ec314451951f/packages/toc/src/generators/notebook/options_manager.ts#L109-L113

krassowski commented on packages/toc/src/generators/notebook/toolbar_generator.tsx:
```diff
@@ -95,6 +95,7 @@ function toolbar(options: OptionsManager, tracker: INotebookTracker) {
             ) as boolean;
             options.initializeOptions(
               numbering || options.numbering,
+              options.numberingH1,
```
Should it be preserved in the notebook metadata as well, so that the setting is remembered when user refreshes the page/re-opens the notebook later? All other toggles work this way. It would require adding:

```ts
const numberingH1 = tracker.currentWidget.model!.metadata.get(
   'toc-numberingH1'
) as boolean;
```

changing this line to:

```ts
numberingH1 || options.numberingH1,
```

And updating the setter to also have `this.notebookMetadata = ['toc-numberingH1', this._numberingH1];` as do other setters:

https://github.com/jupyterlab/jupyterlab/blob/dbdefed9db9332381fe4104bdf53ec314451951f/packages/toc/src/generators/notebook/options_manager.ts#L109-L113

krassowski replied:
Oh, wait there is no toggle for this one, just a setting. Well. so I guess it is fine as is, unless there should be a toggle.

fcollonval reviewed (approved):
LGTM thanks for pushing this through @skyetim

krassowski commented on packages/toc-extension/schema/plugin.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "title": "Table of Contents",
```
```suggestion
  "jupyter.lab.setting-icon": "ui-components:toc",
  "jupyter.lab.setting-icon-label": "Table of Contents",
  "title": "Table of Contents",
```

krassowski reviewed (approved):
It looks good to me. I would consider generalising this to allow users to skip any set of heading levels (H2, H3, etc. and not only H1 only), but this can be done in a future PR for 4.0 or later.

Also I left a suggestion on scheme to add an icon for the settings editor.

skyetim commented:
Sounds great. Thanks for all the help and suggestions in the pr.

joelostblom commented:
This is great! Thanks for adding the option to not number h1 headings @skyetim :raised_hands:
